### PR TITLE
feat: Title support via front matter for ER, state, class, git, and graph diagrams

### LINKED ISSUE #1433: Title support for every diagram
**Is your feature request related to a problem? Please describe.**
Titles are really important when sharing diagrams because they give a strong context about what the person is seeing.  

**Describe the solution you'd like**
It would be amazing if all memaid diagrams supported a standard title or "name", so, when sharing just the diagram, it would always be clear what the diagram is about. 

**Describe alternatives you've considered**
The same way we declare the diagram type as the first instruction, if we could declare the diagram name, it would be really cool. Something like this would also be interesting: graph LR (My Beautiful Graph)

### PR DESCRIPTION
**Pull request by MasonM**
## :bookmark_tabs: Summary

This adds title support for all diagram types currently lacking support for titles (excluding "info" and "error") using [Jekyll-style YAML frontmatter blocks](https://jekyllrb.com/docs/front-matter/). C4C diagrams, pie charts, gantt charts, requirement diagrams, sequence diagrams, and user journey diagrams already have title support.

Resolves #1433

## :straight_ruler: Design Decisions

1. This increases the size of `mermaid.min.js` by 33KB and `mermaid.esm.min.js` by 42KB:
    ```
    Before:
    $ ls packages/mermaid/dist/mermaid*min*js -la
    -rw-rw-r-- 1 masonm masonm 1591427 Nov 19 10:27 packages/mermaid/dist/mermaid.esm.min.mjs
    -rw-rw-r-- 1 masonm masonm 1176130 Nov 19 10:27 packages/mermaid/dist/mermaid.min.js
    
    After:
    $ ls packages/mermaid/dist/mermaid*min*js -la
    -rw-rw-r-- 1 masonm masonm 1633782 Nov 19 10:25 packages/mermaid/dist/mermaid.esm.min.mjs
    -rw-rw-r-- 1 masonm masonm 1208723 Nov 19 10:25 packages/mermaid/dist/mermaid.min.js
    ```
   I avoided using [gray-matter](https://www.npmjs.com/package/gray-matter) or similar packages to minimize the bundle size increase and avoid the maintenance hassles associated with new dependencies. 
2. The parser uses the [failsafe schema](https://yaml.org/spec/1.2.2/#101-failsafe-schema) to keep things simple and avoid the Norway problem. 
3. Only support for titles in frontmatter have been added. Support for other types of metadata should be done after this is merged IMO, since that will require more discussion (as noted by @sidharthv96).
4. I didn't deprecate the existing title syntax for those diagram types that already support titles, since I'm not sure which docs to update, and that would make this PR too big IMO. I think that should be done in a follow-up PR.
5. I added a `titleTopMargin` configuration option for every diagram type for consistency. Making it a global option might make more sense, but that could raise backwards-compatibility concerns (e.g. what if both `gantt.titleTopMargin` and the global `titleTopMargin` are set).

### :clipboard: Tasks

Make sure you

- [X] :book: have read the [contribution guidelines](https://github.com/mermaid-js/mermaid/blob/develop/CONTRIBUTING.md)
- [X] :computer: have added unit/e2e tests (if appropriate)
- [X] :bookmark: targeted `develop` branch

### CODE REVIEW

**Comment by MasonM** [comment]
The failure in the "E2E / build (18.x, 1) (pull_request) " check appears unrelated:
>  Warning: Failed to download action 'https://api.github.com/repos/actions/setup-node/tarball/8c91899e586c5b171469028077307d293428b516'. Error: Response status code does not indicate success: 500 (Internal Server Error).

**Review by aloisklink — COMMENTED** [review]
> The failure in the "E2E / build (18.x, 1) (pull_request) " check appears unrelated

Yep! Just seems to be GitHub's servers being flakey.

This isn't a proper review (I'm not an expert on the rendering code, so will leave that to other reviewers), but from one I've seen, it all looks good! Thanks for adding tests and demo examples!

**Potential improvements**:

- Can you add a `title` example to the `packages/mermaid/src/docs/classDiagram.md`?

**Comment by MasonM** [comment]
Thanks for the review @aloisklink! Extracting the rendering logic to a helper method sounds reasonable, so I went ahead and did that and added tests for it. I also updated the docs for `packages/mermaid/src/docs/classDiagram.md`

File: packages/mermaid/src/utils.ts
Comment (aloisklink) [inline packages/mermaid/src/utils.ts:847]: Thanks for making those changes :)

One last thing, not 100% required, but can you add a quick description for these parameters?

Otherwise, we might get some JSDoc ESLint warnings (or we might get some warnings in the future).

Comment (aloisklink) [reply packages/mermaid/src/utils.ts:847]: Actually, now that https://github.com/mermaid-js/mermaid/pull/3708 is merged, you'll definitely need to change these.

TypeScript `.ts` files now need TSDoc comments, not JSDoc comments (you may need to update your PR from the latest changes from `develop`)

Basically, you'll have to remove the type (since it's already in the function), and add a `-` before the description, e.g.

```ts
 * @param parent - d3 svg object to append title to
 * @param cssClass - CSS class to give the `<title>` element
 * @param titleTopMargin - Pixels between title and top of SVG
 * @param title - Title contents (if blank, this does nothing)
```

Comment (MasonM) [reply packages/mermaid/src/utils.ts]: Okay, done in ccd88c8b2928f09823e7838216f93335f7a496dc

**Review by aloisklink — APPROVED** [review]
Looks good to me!

Again, I'm not super experienced with SVG/d3/rendering code, so another reviewer will need to double-check that.

As a quick example, this is what the new title in the <http://localhost:9000/git.html> demo looks like:

demos/git.html with title

**Comment by knsv** [comment]
Wow Mason! A big job, I appreciate changes that align diagrams. One of the challenges going forward is keeping the diagrams in line.

**Review by knsv — CHANGES_REQUESTED** [review]
Hi @MasonM!

Great job but  we are not there yet! I have reviewed you changes and found some issues. When you introduced the **title** keyword you also unwittingly made it impossible to create a state named title, a flowchart with an id of title/usage of title in a subgraph title or an er diagram with an entity title.

We are very restrictive with changes that would break existing diagrams. It is one thing to introduce a change leading to a developer having update the integration of mermaid on a site but whole other thing with an update that is forcing authors to go in and update the already written documentation.

On possible fix for the issue would be, to be more restrictive with scoping out what a title is, in the lexer. The solutions will probably differ between the different diagrams though. Another avenue could be to use the different states in the parser to know if a title could be expected.

One example of a breaking diagram:
```
erDiagram
    title ||--o{ ORDER : places
    CUSTOMER {
        string title
        string custNumber
        string sector
    }
    ORDER ||--|{ LINE-ITEM : title
    ORDER {
        int orderNumber
        string deliveryAddress
    }
    LINE-ITEM {
        string productCode
        int quantity
        float pricePerUnit
    }
```

Good luck,
Knut

**Comment by MasonM** [comment]
@knsv Thanks for the review! I understand the important of retaining backwards-compatibility, but I'd assumed there was some leeway here, since the changes that introduced `accTitle` in https://github.com/mermaid-js/mermaid/pull/3008/ were also breaking changes, for the same reason as these changes (e.g. it broke states diagram with a state named `accTitle:`). It's unlikely those changes affected anyone, since `accTitle` is an uncommon phrase, but it's still possible.

If the policy is that backwards-incompatible changes are strictly forbidden, then the only option I can see is using a directive, e.g.:
```
graph TD
    %%{ title: 'testing' }%%
    foo --> bar
```

That's cumbersome, but it's guaranteed to be backwards-compatible, and all the diagrams already support directives. I can go ahead and implement that if you want, but if there is some leeway here, then changing `title` to `title:` or using slightly different syntax would probably be preferable. Let me know what you'd like.

**Comment by knsv** [comment]
Hi! Yeah, thats the price of a organically growing syntax. If I started today I would make the syntax a little different so that states in the parser can be applied easier but it is as it is.

I am thinking that restricting the lexer regarding the title could be a good place to start. Right now you use `"title"\s[^#\n;]+` for flowcharts. So this makes the parser interpret title followed by space and anything except # or ; or new line as being a title. Here you could be more strict and force a space and a quote, require an end quote before new line.

Hope this helps

**Comment by MasonM** [comment]
>  I am thinking that restricting the lexer regarding the title could be a good place to start. Right now you use "title"\s[^#\n;]+ for flowcharts. So this makes the parser interpret title followed by space and anything except #, ; or new-line as being a title. 

That's still a breaking change for some diagram types. Consider [this state diagram](https://mermaid.live/edit#pako:eNpFzLsNhTAQRNFW0MRU4JgOSDdZ4TFY8geZdYAQveMXvYlucDQPtuoJh8vUuETdm2Yp05hFS5wEoVYBZmS2rNEP-_yAwA5mCtxIz6A9mUDKO6h2q-tdNjhrnTP66f_vcEHTxfcD0sUpiA), which currently renders as a diagram with the states `title` and `"foo"`, but would instead just show the title "foo" after such a change.

It sounds like you're okay with "minor" breaking changes, but the definition of "minor" is quite subjective. That's perfectly reasonable, but I can't read your mind, and I'd prefer not to guess and spend a lot of time implementing a new syntax, only to be told it won't be accepted because it's too big of a breaking change. Could you please tell me the exact grammar you'd like to see for each diagram type?

**Comment by knsv** [comment]
You are right, that solution will only work for flowcharts and er diagrams. We need a different solution for state diagrams.

I would to start with proposing to delay State diagrams a bit so that we can get this done in October as a part of Hacktoberfest.

**Comment by MasonM** [comment]
@knsv It's going to break class diagrams too. [Example](https://mermaid.live/edit#pako:eNo1jr0OwjAMhF8lurl9gYiRlQlGL1biQKT8oMQZUNV3JxXlFp_us-zb4KoXWLjEvV8jPxtnKmZKoyYxhFArwVzW1Tyk64_9hQVZWubo54ntYAR9SRaCndZL4JGUQGWfqzy03j_FwWobsmC8PaucT2EDpz5T8VFru521jrF_Aap9NNI).

As I see it, there's four options here:
1. Accept there's going to be some breakage and stick with the current syntax, as was done with `accTitle` in https://github.com/mermaid-js/mermaid/pull/3008
2. Go with directives, as I suggested earlier. This has zero chance of breaking anything. The only downside is it's cumbersome.
3. Go with the the `title "test"` syntax you suggested, but only for flowcharts and ER diagrams, and revert all the changes in this PR for all the other diagram types. This means Mermaid will end up having at least three different syntaxes for defining diagrams: `title test`, `test "test"`, and whatever ends up being chosen for  the other diagram types.
4. Close this PR and abandon the whole thing

Which of these 4 options do you want to do?

**Comment by knsv** [comment]
Thanks for your analysis. Some feedback:
1: This would be nice but **title** is a much to common word. Many diagrams will break. I don't think this option is realistic. We cant do this.
2: I agree that this one is safe but I also agree with you that it is cumbersome. I think that no one will understand how to use it.
3. The point is to make the syntax standardised between the diagrams. Then this is not suitable either.
4. I can see that this is growing a bit larger then you perhaps knew when starting out on this journey. I understand if you think this is too complicated in all the wrong ways but the essence of this issue is sound and I think we should solve it. Your help would be appreciated.

Some broader ideas:
* Let's deprecate the current way to set the title in the diagrams that can set the title. In practice, keep it in the parsers indefinitely in the old way but remove it from the documentation.
* These standardised functions should use the common db instead of putting the same function in all diagrams databases.
* Perhaps the syntax could be generic to be utilised for future properties as well, something like this perhaps:
    - `_set: title "The title"`
    - `_set: otherProperty "value"`

What do you think?

**Comment by MasonM** [comment]
> Perhaps the syntax could be generic to be utilised for future properties as well, something like this perhaps: 

That's a good idea. Having special syntax for distinguishing metadata (like the title) from the actual diagram data is likely be useful in the future. Implementing it in a backwards-compatible way will be tricky, but it's only going to get more challenging as Mermaid continues to grow and adds more diagrams and syntax. 

State diagrams are particularly challenging due to the lax grammar, and `_set: title "foo"` won't work since it's currently parsed as states `_set` and `title` ([example](https://mermaid.live/edit#pako:eNpFTr0KAjEYe5WS-VwcO7s6uRbk4_pVC-1V2lSQ497dnjeYKeSHZMVcvMKiUaiXKI8q-fQ-u8UM3JvSGkYmNQ6hFIfDoDZiQtaaJfpRX3fdgU_N6mAH9RqkJ-6NbUSls9w-ywzL2nVCf_n_IGyQ1IaqPrLU63Hp92z7AujANgE)). A possible alternative is just prefixing it with a hyphen (i.e. `-title foo`), since hyphens aren't allowed in state names. From a quick look, it doesn't appear that syntax breaks any of the other diagram types either, but I'll need to spend more time this weekend to check. If it does work, then it could be generalized to the grammar `"-" metadataName "\s+" metadataValue` for representing arbitrary metadata.

**Comment by aloisklink** [comment]
> That's a good idea. Having special syntax for distinguishing metadata (like the title) from the actual diagram data is likely be useful in the future. Implementing it in a backwards-compatible way will be tricky, but it's only going to get more challenging as Mermaid continues to grow and adds more diagrams and syntax.

Hmmmm, what about having an optional YAML header in a mermaid definition?

In many markdown software, you can optionally add a YAML header to add Markdown metadata, e.g. something like:

```markdown
---
title: My title
date: 1970-01-01
author: FirstName LastName
---

# Hello world
```

The rules would be the same as in Markdown, it must be the first thing in a Mermaid definition, it might be valid YAML, and it must be between `---` lines, e.g.

````markdown
Here is my mermaid definition:

```mermaid
---
title: my graph
author: FirstName LastName
---
flowchart LR
    A[Hard] -->|Text| B(Round)
    B --> C{Decision}
```
````

Benefits:
  - YAML is well known to most people that write Markdown
  - YAML has a bunch of helpful features, e.g. the ability to write very long strings, comments, nested arrays/objects
  - If we stick it before the graph definition, we could have a separate pre-processor that runs before the graphs, which would be easier to code.

**Comment by weedySeaDragon** [comment]
Jumping in (since I'm currently working with the state diagram parser) --

Why is `title: ` not acceptable?  I've read through the comments but have somehow missed why that isn't acceptable.

Is it only the problem with the state diagrams?

**Comment by aloisklink** [comment]
> Why is `title:` not acceptable? I've read through the comments but have somehow missed why that isn't acceptable.

I believe it's because of backwards compatibility, there's a pretty good chance somebody has already created a mermaid state diagram that uses `title:` in it, and so changing the behavior of `title:` might break their diagrams.

E.g. something like:

````markdown
```mermaid
stateDiagram
  title: Example state diagram
  my2ndState: My second state
  title --> my2ndState
```
````

```mermaid
stateDiagram
  title
  title: Example state diagram
  my2ndState: My second state
  title --> my2ndState
```

Or

````markdown
```mermaid
stateDiagram
  title
  description
  title --> description
```
````

```mermaid
stateDiagram
  title
  description
  title --> description
```

**Comment by weedySeaDragon** [comment]
Ah Makes total sense.  Thanks for spelling that out for me @aloisklink .

**Comment by MasonM** [comment]
Okay, I couldn't find anything that would break with the `-title foo` syntax, so I went ahead and pushed a commit to switch to that, but I do like @aloisklink's suggestion of using a YAML front matter block. I believe that was pioneered by [Jekyll](https://jekyllrb.com/docs/front-matter/), and is also supported by [Pandoc](https://pandoc.org/MANUAL.html#extension-yaml_metadata_block). It has a few disadvantages, however:
1. The syntax is significantly more verbose than the `"-" metadataName "\s+" metadataValue` syntax, but it's definitely more readable, which is more important IMO.
2. We'd need an extra package to handle parsing the front matter block. There is an NPM module called [front-matter](https://www.npmjs.com/package/front-matter), but it's 171KB minified [according to Bundlephobia](https://bundlephobia.com/package/front-matter@4.0.2), which is probably unacceptable. The [js-yaml](https://www.npmjs.com/package/js-yaml) package is  ~40KB minified, and is already an indirect dev dependency of Mermaid. There was an arbitrary code execution vulnerability in js-yaml v3 a few years ago ([advisory](https://github.com/advisories/GHSA-8j8c-7jfh-h6hx)), which is worrisome, but it appears v4 made a lot of changes to improve security.
3. Directives are in JSON format, so having front matter in YAML could be a little confusing

Personally, I think YAML is probably the right way to go. If @knsv agrees, then I can go ahead and implement that.

**Comment by knsv** [comment]
I am all for it!

Further down the line if things work out I could see directives pulled in to this syntax. The old syntax could remain for backwards compatibility, this would be the documented one. The size increase is worry-some as mermaid is on the heavy side already.

I am convinced that having a preprocessing step is the method to use. This will eliminate conflicts with the diagrams.

**Comment by AssassinLV** [comment]
Why not then just:

````
```mermaid
title: my graph
author: FirstName LastName
flowchart LR
    A[Hard] -->|Text| B(Round)
    B --> C{Decision}
```
````

without encapsulating in `---` as it would still be above the actual diagram definition.

**Comment by aloisklink** [comment]
> Directives are in JSON format, so having front matter in YAML could be a little confusing

JSON is valid YAML, so it should be easy to migrate directives to YAML, if we want to support that.

> without encapsulating in --- as it would still be above the actual diagram definition

It makes sense if we're using YAML for a few reasons:
  - Everybody else does it for YAML headers (e.g. https://docusaurus.io/docs/markdown-features#front-matter)
  - `---` is part of the YAML and [RFC 2046](https://datatracker.ietf.org/doc/html/rfc2046) spec, and I believe you can tell a YAML parser to just scan the bit between `---`s (not 100% sure).
  - It's easier to code that way, we can just say that if a mermaid definition starts with a `---`, run the YAML pre-parser, else continue.
    - It will also make syntax highlighting for mermaid code easier!

Basically, using `---` and YAML should hopefully mean less code to write!

**Comment by aloisklink** [comment]
> We'll have to draft a spec for what goes into the yaml.

It might be worth specifying this using https://json-schema.org/, since YAML can almost always be converted to JSON (I think the only issue is with circular references, but those are very rare).

We could then use https://www.npmjs.com/package/json-schema-to-typescript to easily convert these into TypeScript types.

And we could use those JSON Schemas for input validation, and print good errors if something is invalid (e.g. what Webpack does).

If I have time, I'm considering moving the Mermaid config to JSON Schema too, since the way it is currently documented is incompatible with the new Vitepress website.

**Comment by MasonM** [comment]
Okay, I implemented frontmatter handling and changed all the diagrams there were previously modified in this PR to use it for titles. Notes:
1. This increases the size of `mermaid.min.js` by 41KB and `mermaid.esm.min.js` by 54KB. Like I mentioned earlier, I avoided using [gray-matter](https://www.npmjs.com/package/gray-matter) or similar packages to minimize the bundle size increase and avoid the maintenance hassles associated with new dependencies. 
2. The parser uses the [failsafe schema](https://yaml.org/spec/1.2.2/#101-failsafe-schema) to keep things simple and avoid the Norway problem. 
3. Only support for titles in frontmatter have been added. Support for other types of metadata should be done after this is merged IMO, since that will require more discussion (as noted by @sidharthv96), and I think this PR is too big already.
4. I didn't deprecate the existing title syntax for those diagram types that already support titles, since I'm not sure which docs to update, and this PR is getting too big. I think that should be done in a follow-up PR.

**Review by sidharthv96 — APPROVED** [review]

File: packages/mermaid/src/diagram-api/frontmatter.ts
Comment (aloisklink) [inline packages/mermaid/src/diagram-api/frontmatter.ts:22]: We should probably give `db` a type, even if it is a very basic one.

Could we do something like:

```ts
/**
 * Generic Diagram DB that may apply to any diagram type.
 */
interface DiagramDb {
    setDiagramTitle?: (title: string) => void
};
```

Ideally, we'd use classes/inheritance for the databases, but that will be a big refactor for the future.

**Review by aloisklink — CHANGES_REQUESTED** [review]
Looks amazing! And you've made this PR pretty small too, which means it's a pretty good change!

Some minor nit-picks from me, but they're mostly just nit-picks.

The only required change is to add a test for a YAML that contains `---` on it's own line, e.g.:

```yaml
title: |
    My multi-line string
    ---
    ^^^ Has an end-document string in it in it's own line
```

It's a rare edge-case, but we might run into it if we make a flowchart `mermaid` and we want to add a YAML front-matter example to a description :)

We should also change the PR title to saying something about YAML frontmatters. E.g. something like: `Add YAML frontmatter to set titles in diagrams`. It might also be worth copying some of https://github.com/mermaid-js/mermaid/pull/3706#issuecomment-1320797780 to the main PR description with a quick example.

File: packages/mermaid/src/diagram-api/frontmatter.spec.ts
Comment (MasonM) [reply packages/mermaid/src/diagram-api/frontmatter.spec.ts:42]: I don't see any way of supporting that syntax without serious compromises:
1. Disallowing leading whitespace before the delimiters would work (that's what gray-matter does), but leading whitespace before a diagram is very common, and disallowing it means you can't indent a diagram. e.g. https://github.com/mermaid-js/mermaid/blob/70f024735b57ac5b29b5c339ac871078ab46fa5a/demos/flowchart.html#L19-L24
2. Changing the `(.*?)` to `(.*)` isn't an option because that will break compatibility with diagrams containing `---`
3.  Trying to match the indent level of the closing delimiter with that of the opening delimiter won't work because leading whitespace is stripped in https://github.com/mermaid-js/mermaid/blob/70f024735b57ac5b29b5c339ac871078ab46fa5a/packages/mermaid/src/mermaid.ts#L149-L153

Do you have any other ideas? If not, I think we should consider this an edge case and not support it

Comment (aloisklink) [reply packages/mermaid/src/diagram-api/frontmatter.spec.ts:42]: > Disallowing leading whitespace before the delimiters would work

My gut feeling is that this is the approach we should go for (as that's that is what [YAML spec v1.0 section 4.3.1](https://yaml.org/spec/1.0/#syntax-stream-doc) says), but you are right that many people do indent the diagrams.

We could add a [`dedent`](https://www.npmjs.com/package/ts-dedent) to https://github.com/mermaid-js/mermaid/blob/70f024735b57ac5b29b5c339ac871078ab46fa5a/packages/mermaid/src/mermaid.ts#LL149-L153. It would only apply to HTML however.

We could also add `dedent` inside `parse(str)` somewhere (or other functions that takes the Mermaid string directly), e.g. right before:

https://github.com/mermaid-js/mermaid/blob/09ed41b7d2cd5c8ffd50c7544bc88bd213d3c601/packages/mermaid/src/mermaidAPI.ts#L81

But that's not super required. People calling the JavaScript API can use `dedent` themselves if they want to use indented strings and the YAML frontmatter, e.g.

```js
import { dedent } from 'ts-dedent';
const myDiagram = dedent`
  ---
  title: This YAML is indented by two spaces.
  ---
  flowchart
    "This diagram is indented by two spaces."
`;

const svg = mermaid.render("mermaid-render-element", myDiagram);
```

Comment (MasonM) [reply packages/mermaid/src/diagram-api/frontmatter.spec.ts:42]: Well, alright, if you're sure that's the way to go, I implemented that in 7eba4f930bf56d8811ab00c93ad0b875dbd4c4c0

File: packages/mermaid/src/diagram-api/frontmatter.ts
Comment (MasonM) [reply packages/mermaid/src/diagram-api/frontmatter.ts]: Done in 3316aa5f4f8c93f3410ca74c2fa5390087f26ef9

File: packages/mermaid/src/diagram-api/frontmatter.ts
Comment (MasonM) [reply packages/mermaid/src/diagram-api/frontmatter.ts]: Done in 3316aa5f4f8c93f3410ca74c2fa5390087f26ef9

File: packages/mermaid/src/diagram-api/frontmatter.ts
Comment (MasonM) [reply packages/mermaid/src/diagram-api/frontmatter.ts]: Done in 3316aa5f4f8c93f3410ca74c2fa5390087f26ef9

File: packages/mermaid/src/diagram-api/frontmatter.spec.ts
Comment (aloisklink) [reply packages/mermaid/src/diagram-api/frontmatter.spec.ts:42]: > if you're sure that's the way to go

I'm not fully sure! But I'd rather have the parser be too strict at the the beginning, since we can always loosen things later without it being a breaking change. It's much harder to make something stricter later on, as that would be a breaking change.

But I'd be be happy for people to disagree or if they think there's a better way to do things!

I am slightly leaning towards also adding a `dedent` before parsing the Mermaid HTML in [mermaid.init/mermaid.initThrowsErrors](https://github.com/mermaid-js/mermaid/blob/70f024735b57ac5b29b5c339ac871078ab46fa5a/packages/mermaid/src/mermaid.ts#L149-L153) so that Mermaid diagrams indented in HTML work, but not for JS code in `mermaid.parse`.

But that could be a decision for a future PR :)

**Review by aloisklink — APPROVED** [review]
I'm happy with this!

For other reviewers, currently, the YAML must **not** be indented, e.g.:

```html
<!-- This is fine, no indentation on the YAML --!>
<pre class="mermaid">
---
title: My title
---
flowchart
  my flowchart
</pre>

<!-- This is not fine, there is indentation on the YAML --!>
<pre class="mermaid">
    ---
    title: My title
    ---
    flowchart
        my flowchart
</pre>
```

It may be worth using [dedent](https://www.npmjs.com/package/ts-dedent ) to automatically remove the indentation from Mermaid code in HTML, but I'm happy for that to be added in a different PR. See discussion in https://github.com/mermaid-js/mermaid/pull/3706#discussion_r1027127998 for more info.

**Comment by MasonM** [comment]
@knsv Could you please review this? Unfortunately, I have very limited additional free time to spend on this, so if this isn't merged soon, I'll have to close it.

**Comment by knsv** [comment]
Will review today

File: packages/mermaid/src/diagram-api/frontmatter.spec.ts
Comment (knsv) [reply packages/mermaid/src/diagram-api/frontmatter.spec.ts:42]: When adding diagrams in html you inherit the indentation from the tag. This looks better in HTML:
```
    <pre id="diagram" class="mermaid">
      ---
      title: Example Git diagram
      ---
      gitGraph
        commit
        commit
    </pre>
```

than:
```
    <pre id="diagram" class="mermaid">
---
title: Example Git diagram
---
gitGraph
  commit
  commit
    </pre>
```

Even though it is less compliant with the standard.

I don't see the multi-line issue as a big problem though. My guess is that the user impact of indenting will affect far, far more users then the lack of ability to use `---` in the beginning of a a multiline row. 

I lean towards the more loose application, better to let the very few users that wants to have the --- in the multiline comment fail instead of irritating the users that want to indent.

But Alois has a fair point. Once we let the cat out of the box, it is out. Lets start with this but I suspect that we need to discuss this more and that we might might want to reverse the a11ab3d5ea1477c02101eb7c4279b1dc0d2ef126 commit. That can be outside this PR though. Good job @MasonM Solid!

**Comment by knsv** [comment]
Approving this but lets add dedent. I like that idea. Solves the indentation issue that will annoy some users!

**Review by knsv — APPROVED** [review]

**Pull request merged**
